stdlib/public/core/Diffing.swift

```diff
@@ -0,0 +1,782 @@
+    var result = Self()
+    var enumeratedRemoves = 0
+    var enumeratedInserts = 0
+    var enumeratedOriginals = 0
+    var currentIndex = self.startIndex
+
+    func append(into target: inout Self, contentsOf source: Self, from index: inout Self.Index, count: Int) {
+      let start = index
+      source.formIndex(&index, offsetBy: count)
+      target.append(contentsOf: source[start..<index])
+    }
+
+    Self.fastApplicationEnumeration(of: difference) { change in
+      switch change {
+      case .remove(offset: let offset, element: _, associatedWith: _):
+        let origCount = offset - enumeratedOriginals
+        append(into: &result, contentsOf: self, from: &currentIndex, count: origCount)
+        enumeratedOriginals += origCount + 1 // Removal consumes an original element
+        currentIndex = self.index(after: currentIndex)
+        enumeratedRemoves += 1
+      case .insert(offset: let offset, element: let element, associatedWith: _):
+        let origCount = (offset + enumeratedRemoves - enumeratedInserts) - enumeratedOriginals
+        append(into: &result, contentsOf: self, from: &currentIndex, count: origCount)
+        result.append(element)
+        enumeratedOriginals += origCount
+        enumeratedInserts += 1
+      }
+      assert(enumeratedOriginals <= self.count)
+    }
+    let origCount = self.count - enumeratedOriginals
+    append(into: &result, contentsOf: self, from: &currentIndex, count: origCount)
+
+    assert(currentIndex == self.endIndex)
+    assert(enumeratedOriginals + origCount == self.count)
+    assert(result.count == self.count + enumeratedInserts - enumeratedRemoves)
+    return result
+  }
+}
+
+// MARK: Definition of API
+
+@available(swift, introduced: 5.1)
+extension BidirectionalCollection {
+  /// Returns the difference needed to produce the receiver's state from the
+  /// parameter's state, using the provided closure to establish equivalence
+  /// between elements.
+  ///
+  /// This function does not infer moves.
+  ///
+  /// - Parameters:
+  ///   - other: The base state.
+  ///   - areEquivalent: A closure that returns whether the two
+  ///     parameters are equivalent.
+  ///
+  /// - Returns: The difference needed to produce the reciever's state from
+  ///   the parameter's state.
+  ///
+  /// - Complexity: For pathological inputs, worst case performance is
+  ///   O(`self.count` * `other.count`). Faster execution can be expected
+  ///   when the collections share many common elements.
+  public func difference<C>(
+    from other: C, by areEquivalent: (Element, C.Element) -> Bool
+  ) -> CollectionDifference<Element>
+    where C : BidirectionalCollection, C.Element == Self.Element
+  {
+    var rawChanges: [CollectionDifference<Element>.Change] = []
+
+    let source = CountingIndexCollection(other)
+    let target = CountingIndexCollection(self)
+    let result = CollectionChanges(from: source, to: target, by: areEquivalent)
```

Comment:
Is `result` really a good name here? Perhaps it would be okay to just generate the changes directly in the `for` statement:

```swift
    for change in CollectionChanges(from: source, to: target, by: areEquivalent) {
```